Check for runs of zero-bytes when linking AssemblerBuffer on x86

Achierius opened the pull request:
#### 207db7247e05a68f966c09dbd67c8a0a2ec5f522
<pre>
Check for runs of zero-bytes when linking AssemblerBuffer on x86
<a href="https://bugs.webkit.org/show_bug.cgi?id=276913">https://bugs.webkit.org/show_bug.cgi?id=276913</a>
<a href="https://rdar.apple.com/132273853">rdar://132273853</a>

Reviewed by Yijia Huang.

The goal here is to detect memory corruption of the AssemblerBuffer
prior to link-time. On ARM64E, these would be caught at link-time via
the ARM64EHash feature within AssemblerBuffer, but due to that feature
being ARM64-specific, on x86 the link would succeed, leaving the
corruption to surface only when the code is executed, which makes it
much more difficult to identify the root cause.

* Source/JavaScriptCore/jit/ExecutableAllocator.h:
(JSC::performJITMemcpy):
* Source/WTF/wtf/PlatformEnable.h:

Canonical link: <a href="https://commits.webkit.org/282011@main">https://commits.webkit.org/282011@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b66271b50649cbdaf039e0ac321a2ab274e228d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61744 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14336 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65723 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12289 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48784 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12560 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8535 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64813 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38180 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53485 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/30633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34837 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10715 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11220 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54839 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/56642 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11018 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67451 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60985 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5687 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53432 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/57404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4682 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82747 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9303 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14473 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37982 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39078 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/JavaScriptCore/assembler/AssemblerBuffer.h:
```diff
@@ -439,6 +439,7 @@ namespace JSC {
 #if !CPU(ARM64) // If we were to define this on arm64e, we'd need a way to update the hash as we write directly into the buffer.
         void* data() const { return m_storage.buffer(); }
 #endif
+        const void* constData() const { return m_storage.buffer(); }
```
This is ugly but unfortunately the `void* data() const` above means we can't also have a `const void* data() const`.

danlliu commented on Source/JavaScriptCore/assembler/LinkBuffer.cpp:
```diff
@@ -472,6 +473,48 @@ void LinkBuffer::linkCode(MacroAssembler& macroAssembler, JITCompilationEffort e
+        size_t runLength = 0;
+        size_t i = 0;
+        if (codeSize > stride) {
+            for (; i % stride; i++) {
```
If `i` is always initialized to zero, is this loop necessary? `i % stride` would end up as `0` immediately, meaning that we'd just exit the loop (unless I'm missing something?).

MenloDorian commented on Source/WTF/wtf/PlatformEnable.h:
```diff
@@ -926,6 +926,10 @@
 #define ENABLE_JIT_SIGN_ASSEMBLER_BUFFER 1
 #endif
 
+#if !defined(ENABLE_JIT_SCAN_ASSEMBLER_BUFFER_FOR_ZEROES) && CPU(X86_64)
```
Let's add ` && PLATFORM(MAC)` as well.  No need to impact other ports.

hyjorc1 reviewed (approved):
r=me

hyjorc1 commented on Source/JavaScriptCore/jit/ExecutableAllocator.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include <bit>
```
I think we don't need this right? `bit_cast` was replaced with `reinterpret_cast`.

Achierius replied:
It's used for std::countl_zero and std::countr_zero

Achierius committed:
Check for runs of zero-bytes when linking AssemblerBuffer on x86
https://bugs.webkit.org/show_bug.cgi?id=276913
rdar://132273853

Reviewed by Yijia Huang.

The goal here is to detect memory corruption of the AssemblerBuffer
prior to link-time. On ARM64E, these would be caught at link-time via
the ARM64EHash feature within AssemblerBuffer, but due to that feature
being ARM64-specific, on x86 the link would succeed, leaving the
corruption to surface only when the code is executed, which makes it
much more difficult to identify the root cause.

* Source/JavaScriptCore/jit/ExecutableAllocator.h:
(JSC::performJITMemcpy):
* Source/WTF/wtf/PlatformEnable.h:

Canonical link: https://commits.webkit.org/282011@main
